nens/G4AW-lizard-client
# [DONE] Offline support

What this PR does (technically)
=============

Leverages the Service Worker Cache API to make this app follow the _Application Shell architecture_.

- Uses https://github.com/NekR/offline-plugin/ for Webpack.
- See https://github.com/nens/G4AW-lizard-client/issues/123 for related issue.


What this PR adds (for the user)
==============

This allows the user to get the application bundle just once.
After this initial visit, even without an internet connection, the application will load itself and the only data over the wire will be HTTP request/responses for the API endpoints and map tiles.


How to test
=========

Use a smartphone, visit the app, close the tab, turn on Airplane mode, open a new tab, visit the app again. It should load the application shell.


How it works
==========

- Webpack is configured to use the `OfflinePlugin` via the `plugins` section.
- OfflinePlugin takes care of generating a Cache Manifest [0].
- Clients look at the cache manifest to see if they need to invalidate their SerrviceWorker (SW) cache.
- The `externals` part of the configuration tells the SW to cache files that are not part of the webpack build, such as the Iconfont and Leaflet CSS. 



References
========

**[0]** The cache manifest looks like this:

```
CACHE MANIFEST
#ver:7/20/2017, 12:12:31 PM
#plugin:4.8.3

CACHE:
/scripts/G4AWLizardClient.js
/
/scripts/../leaflet/leaflet.css
/scripts/../iconfont/material-icons.css
/scripts/../iconfont/MaterialIcons-Regular.eot
/scripts/../iconfont/MaterialIcons-Regular.ijmap
/scripts/../iconfont/MaterialIcons-Regular.svg
/scripts/../iconfont/MaterialIcons-Regular.ttf
/scripts/../iconfont/MaterialIcons-Regular.woff
/scripts/../iconfont/MaterialIcons-Regular.woff2
/scripts/../iconfont/codepoints
/scripts/../iconfont/material-icons.css

NETWORK:
*
```
